api/src/main/java/org/openmrs/module/reporting/cohort/definition/library/BuiltInCohortDefinitionLibrary.java

```diff
@@ -103,6 +105,19 @@ public CohortDefinition getAnyEncounterDuringPeriod() {
         return new MappedParametersCohortDefinition(cd, "onOrAfter", "startDate", "onOrBefore", "endDate");
     }
 
+    @DocumentedDefinition("anyEncounterDuringPeriodWithOccurrence")
+    public CohortDefinition getAnyEncounterDuringPeriodWithOccurrence() {
+        EncounterCohortDefinition cd = new EncounterCohortDefinition();
+        cd.addParameter(new Parameter("onOrAfter", "reporting.parameter.startDate", Date.class));
+        cd.addParameter(new Parameter("onOrBefore", "reporting.parameter.endDate", Date.class));
+        cd.addParameter(new Parameter("atLeastCount", "reporting.parameter.atleast", Integer.class));
+        cd.addParameter(new Parameter("atMostCount", "reporting.parameter.atMost", Integer.class));
+        cd.addParameter(new Parameter("encounterTypeList", "reporting.parameter.encounterTypeList", EncounterType.class, List.class, null));
+        cd.addParameter(new Parameter("formList", "reporting.parameter.formList", Form.class, List.class, null));
+        cd.addParameter(new Parameter("locationList", "reporting.parameter.location", Location.class, List.class, null));
+        return new MappedParametersCohortDefinition(cd, "onOrAfter", "fromDate", "onOrBefore", "toDate", "encounterTypeList", "encounterTypes", "locationList", "location", "atLeastCount", "atLeast", "atMostCount", "atMost", "formList", "form");
```

Comment:
I think we should prioritize consistency and clarity for the consumer of this API.

So, definitely we should consistently call things "startDate" and "endDate" when it's part of a definition with "period" in its name.

I also prefer to shorten things that are called "list" in the Java API to just give them plural names in the REST API. E.g. in "anyEncounterOfTypesDuringPeriod" I called it "encounterTypes" instead of "encounterTypeList", and I would vote for maintaining this trend. (@mseaton, any objections to that?)

The name of this definition is misleading. If there is something called "anyEncounterDuringPeriodWithOccurrence" I would assume that it supports (only) start/end dates and min/max occurrences, and "any" doesn't fit with the "occurrences".

I suggest we call this something like "advancedEncounterSearch" or even just "encounterCohortQuery" and have it expose all of the parameters on the EncounterCohortDefinition class. (This leaves the door open to adding more parameters in the future, without worrying about the name being misleading.)